### PR TITLE
kvserver: extract writeUnreplicatedSST

### PR DESCRIPTION
**Pull request by tbg**
This PR extracts a standalone method `writeUnreplicatedSST` that can be moved
to `kvstorage` at an appropriate time.

This is a sister PR to #96871, which does the same to some other SSTs created
in `applySnapshot`.

Epic: CRDB-220
Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_raftstorage.go [hunk 1]
```diff
@@ -482,59 +482,64 @@ func (r *Replica) applySnapshot(
...
-		return err
+		if err := unreplicatedSST.Finish(); err != nil {
+			return nil, false, err
+		}
+		return unreplicatedSSTFile, unreplicatedSST.DataSize > 0, nil
 	}
-	if unreplicatedSST.DataSize > 0 {
+	unreplicatedSSTFile, nonempty, err := writeUnreplicatedSST(
+		ctx, r.ID(), r.ClusterSettings(), nonemptySnap.Metadata, hs, &r.raftMu.stateLoader.StateLoader,
+	)
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_raftstorage.go:487]: what about checking err != nil?

File: pkg/kv/kvserver/replica_raftstorage.go [hunk 2]
```diff
@@ -704,6 +658,60 @@ func (r *Replica) applySnapshot(
...
 }
 
+func writeUnreplicatedSST(
+	ctx context.Context,
+	id storage.FullReplicaID,
+	st *cluster.Settings,
+	meta raftpb.SnapshotMetadata,
+	hs raftpb.HardState,
+	sl *logstore.StateLoader,
+) (*storage.MemFile, bool, error) {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_raftstorage.go:668]: Please add a func comment, and/or name the output variables so that the returned `bool` is explained.

**Review by pav-kv — APPROVED** [review]

File: pkg/kv/kvserver/replica_raftstorage.go (hunk 1 above)
Comment (tbg) [reply pkg/kv/kvserver/replica_raftstorage.go:487]: Good catch - that CI didn't catch this is surprising[^1]

[^1]: https://cockroachlabs.slack.com/archives/C023S0V4YEB/p1676290099032059

File: pkg/kv/kvserver/replica_raftstorage.go (hunk 2 above)
Comment (tbg) [reply pkg/kv/kvserver/replica_raftstorage.go]: Done. I also added a TODO to look into this parameter in the first place; won't it always be true? Once we have a datadriven test we'll know for sure.

**Review by tbg — COMMENTED** [review]
TFTR!

bors r=pavelkalinnikov

**Pull request merged**